src/biogeochem/Fan2CTSMMod.F90

```diff
@@ -0,0 +1,1059 @@
+    ! N excreted by animals allocated to mixed production systems temporarily grazing on grasslands:
+    real(r8), intent(inout) :: n_manure_graze_col(bounds%begc:bounds%endc)
+    ! N excreted by animals in mixed systems, total
+    real(r8), intent(out) :: n_manure_mixed_col(bounds%begc:bounds%endc)
+    ! NH3 emission fluxes from manure storage and housings, gN/m2/s
+    real(r8), intent(out) :: nh3_flux_stores(bounds%begc:bounds%endc), nh3_flux_barns(bounds%begc:bounds%endc)
+    ! total nitrogen flux transferred out of a crop column (manure spreading + temporary grazing)
+    real(r8), intent(out) :: man_n_transf(bounds%begc:bounds%endc)
+    ! Total nitrogen excreted in barns
+    real(r8), intent(out) :: man_n_barns(bounds%begc:bounds%endc)
+    ! fraction of manure excreted when grazing
+    real(r8), intent(out) :: grz_fract(bounds%begc:bounds%endc)
+    ! TAN fraction in excreted N
+    real(r8), intent(in) :: tan_fract_excr
+    integer                  , intent(in)    :: num_soilc       ! number of soil columns in filter
+    integer                  , intent(in)    :: filter_soilc(:) ! filter for soil columns
+
+    integer :: begg, endg, g, l, c, il, counter, col_grass, status, p
+    real(r8) :: flux_avail_rum, flux_avail_mg, flux_grazing, invscale
+    real(r8) :: tempr_ave, windspeed_ave ! windspeed and temperature averaged over agricultural patches
+    real(r8) :: tempr_barns, tempr_stores, vent_barns, flux_grass_crop, tempr_min_10day, &
+         flux_grass_graze, flux_grass_spread, flux_grass_spread_tan, flux_grass_crop_tan
+    real(r8) :: cumflux, totalinput, total_to_store, total_to_store_tan
+    ! dimensions are (type of flux, ruminant/other)
+    real(r8) :: fluxes_nitr(num_fluxes,2), fluxes_tan(num_fluxes,2) 
+    ! The fraction of manure applied continuously on grasslands (if present in the gridcell)
+    real(r8), parameter :: kg_to_g = 1e3_r8
+    logical :: is_grass
+    
+    begg = bounds%begg; endg = bounds%endg
+    nh3_flux_stores(bounds%begc:bounds%endc) = 0_r8
+    nh3_flux_barns(bounds%begc:bounds%endc) = 0_r8
+    man_n_barns(bounds%begc:bounds%endc) = 0.0_r8
+    
+    totalinput = 0.0
+    cumflux = 0.0
+    
+    do g = begg, endg
+       ! First find out if there are grasslands in this cell. If yes, a fraction of
+       ! manure can be diverted to them before storage.
+       col_grass = ispval
+       do il = 1, max_lunit
+          l = grc%landunit_indices(il, g)
+          if (l == ispval) cycle
+          if (lun%itype(l) == istsoil) then
+             do p = lun%patchi(l), lun%patchf(l)
+                is_grass = patch%itype(p) == nc4_grass &
+                     .or. patch%itype(p) == nc3_nonarctic_grass &
+                     .or. patch%itype(p) == nc3_arctic_grass
+                if (is_grass .and. col%wtgcell(patch%column(p)) > 1e-6) then
+                   col_grass = patch%column(p)
+                   exit
```

Comment:
Here is another instance where it seems like the answers will depend on the order of different patches in memory (at least in the case where each patch is on its own column).

Reply:
I did not know that the natural pfts could have several columns. I changed this to take the istsoil column from the landunit, which is much simpler anyway.

Reply:
The new logic that checks `if (col%itype(c) == istsoil)` is problematic. istsoil is used for landunits, not column types. It happens to work here because it happens that we use a column type of 1 for the natural vegetated column, but that isn't guaranteed. Furthermore, the new logic won't do what you want if there are actually multiple vegetated columns in a grid cell. As a short-term fix, I might suggest returning to something closer to the old logic, but we should flag this as something to come back to. 

**This one is important to fix (unlike most of my other outstanding comments).**

Reply:
OK, maybe I misunderstood the comment on the top of the ColumnType module. Can you describe in more detail the configuration that this logic should support? Is the idea to have the istsoil landunit contain several columns with the same itype? In this case, will the pfts be unique within the landunit or can the same pft be in multiple columns that differ in some other way?

What I could in principle do is to search sequentially, in separate loops, for each grass pft type, although it feels like an annoyingly complicated solution. But if the same pft can be in several columns within the landunit, then even this won't help.

Reply:
> OK, maybe I misunderstood the comment on the top of the ColumnType module.

No, that was my mistake: I forgot that we documented that.

Even so, I think we'll need to give some more thought to this logic. In the medium-to-long-term, we want to allow a more flexible arrangement of columns on the vegetated landunit, to support hillslope hydrology and other applications. I'm actually not sure at this point whether all columns would have the same column type or different column types, but I think they may need different column types for some subtle reasons. And I think it will be possible for the same pft to be on multiple columns in the landunit.

Your current logic doesn't guarantee that you have found a column with grass PFTs on it, which is why I suggested reverting to your previous logic here. It feels like the right logic long-term might involve allocating the manure to all columns in the grid cell that have grassland on them, maybe proportional to the amount of grass on them. This is complicated, but we may need something like that to avoid having answers change if you (for example) add a zero-weight column to a given grid cell for diagnostic purposes.

But I don't have a full understanding of the thinking behind this code, and I think this warrants a bigger discussion to make sure this will do what's really intended. This also raises some scientific questions related to scale: My understanding is that this code transfers manure from the crop column to the natural veg column on a given grid cell. But allowing for this transfer within a grid cell and not between grid cells feels very scale-dependent: What if you're running at high resolution (like 1 km) with only a single veg type in each grid cell, for example? Would this lead to higher application rates on the crop simply because there isn't any natural veg in the grid cell? Maybe that's okay, but I want to make sure people are comfortable with this scale dependence....

Reply:
> In hindsight, the whole thing should have been done differently. A loop could go over the gridcell level N inputs and partition them between one or more columns in each cell. This could be done according to the same rules that now transfer N from crop to non-crop columns, but there would then be no need for the awkward column-to-column transfers.

Yes, this sounds like a better solution, and probably what we want to do long-term.